[WIP] RHDEVDOCS-3917 - documents new authentication methods for remote_write in monitoring (#44713)
openshift/openshift-docs: modules/monitoring-configuring-remote-write-storage.adoc

* You have installed the OpenShift CLI (`oc`).
* You have set up a remote write compatible endpoint (such as Thanos) and know the endpoint URL.
See the link:https://prometheus.io/docs/operating/integrations/#remote-endpoints-and-storage[Prometheus remote endpoints and storage documentation] for information about endpoints that are compatible with the remote write feature.

> * You have set up authentication credentials in a `Secret` object for the remote write endpoint.

Author: Added this content to clarify the namespace for the secret needs to be the same as the Prometheus object

 
+
[CAUTION]
====
To reduce security risks, avoid sending metrics to an endpoint using unencrypted HTTP or without using authentication.
====

.Procedure

Follow these steps to configure remote write for default platform monitoring in the `cluster-monitoring-config` config map in the `openshift-monitoring` namespace.

[NOTE]
====

> If you want to configure remote write for the Prometheus instance that monitors user-defined projects, make similar edits to the `user-workload-monitoring-config` config map in the `openshift-user-workload-monitoring` namespace

Author: Added this content back and put it at the end of the note admonition.

* You have installed the OpenShift CLI (`oc`).
* You have set up a remote write compatible endpoint (such as Thanos) and know the endpoint URL.
See the link:https://prometheus.io/docs/operating/integrations/#remote-endpoints-and-storage[Prometheus remote endpoints and storage documentation] for information about endpoints that are compatible with the remote write feature.
* You have set up authentication credentials in a `Secret` object for the remote write endpoint.

> The secret needs to be created in the same namespace as the Prometheus object for which you configure remote write: `openshift-monitoring` for default platform monitoring or `openshift-user-workload-monitoring` for user workload monitoring.

Reviewer: ```suggestion
["You must create the secret" OR "The secret must be"] in the same namespace as the Prometheus object for which you configure remote write: `openshift-monitoring` for default platform monitoring or `openshift-user-workload-monitoring` for user workload monitoring.
```

Reviewer: I wonder if we should unpack the meaning a bit to make it easier for non-fluent readers. Perhaps:
```suggestion
The secret must be in the same namespace as the Prometheus object for which you configure remote write. That is, for default platform monitoring, the secret must be in the `openshift-monitoring` namespace. Or, for user workload monitoring, the secret must be in the `openshift-user-workload-monitoring` namespace.
```

 
+
[CAUTION]
====

> To reduce security risks, avoid sending metrics to an endpoint using unencrypted HTTP or without using authentication.

Reviewer: Maybe
```suggestion
To reduce security risks, use HTTPS and authentication to send metrics to an endpoint.
```

Author: revised per your suggestion

====

.Procedure

Follow these steps to configure remote write for default platform monitoring in the `cluster-monitoring-config` config map in the `openshift-monitoring` namespace.

[NOTE]
====

> If you want to configure remote write for the Prometheus instance that monitors user-defined projects, make similar edits to the `user-workload-monitoring-config` config map in the `openshift-user-workload-monitoring` namespace. 

Reviewer: ```suggestion
If you configure remote write for the Prometheus instance that monitors user-defined projects, make similar edits to the `user-workload-monitoring-config` config map in the `openshift-user-workload-monitoring` namespace. 
```

Author: revised per your suggestion

> If you do so, note that the Prometheus config map component is called `prometheusK8s` in the `cluster-monitoring-config` `ConfigMap` object and `prometheus` in the `user-workload-monitoring-config` `ConfigMap` object.

Reviewer: ```suggestion
If you do so, note that the Prometheus config map component is called `prometheusK8s` in the `cluster-monitoring-config` config map and `prometheus` in the `user-workload-monitoring-config` config map.
```

Author: Revised to read `Note that the Prometheus config map component is called `prometheus` in the `user-workload-monitoring-config` `ConfigMap` object and not `prometheusK8s`, as it is in the `cluster-monitoring-config` `ConfigMap` object.`

    prometheusK8s:
      remoteWrite:
      - url: "https://remote-write.endpoint"
        <endpoint_authentication_credentials> <1>
----
+
<1> For `endpoint_authentication_credentials` substitute the authentication method and the credentials for the endpoint.

> Currently supported authentication methods are AWS Signature version 4 (`sigv4`), basic authentication (`basicAuth`),  bearer token (`Bearer`), TLS client (`tlsConfig`), and OAuth 2.0 (`oauth2`). 

Reviewer 2: There is the field `bearerTokenFile` to specify the bearer token. The file should be mounted to the Prometheus pods.
https://github.com/openshift/cluster-monitoring-operator/blob/release-4.11/pkg/manifests/config.go#L158

> Currently supported authentication methods are AWS Signature version 4 (`sigv4`), basic authentication (`basicAuth`),  bearer token (`Bearer`), TLS client (`tlsConfig`), and OAuth 2.0 (`oauth2`). 

Author: Changed all occurrences of `Bearer` to `bearerTokenFile`

> Currently supported authentication methods are AWS Signature version 4 (`sigv4`), basic authentication (`basicAuth`),  bearer token (`Bearer`), TLS client (`tlsConfig`), and OAuth 2.0 (`oauth2`). 

Reviewer 3: lgtm

> Currently supported authentication methods are AWS Signature version 4 (`sigv4`), basic authentication (`basicAuth`),  bearer token (`Bearer`), TLS client (`tlsConfig`), and OAuth 2.0 (`oauth2`). 

Reviewer 3: the doc is for epic https://issues.redhat.com/browse/MON-2160, there is one task
https://issues.redhat.com/browse/MON-2207 Expose Authorization settings for remote write in the CMO configuration
`authorization` is defined in
https://github.com/openshift/cluster-monitoring-operator/blob/release-4.11/pkg/manifests/config.go#L160
`Authorization *monv1.SafeAuthorization `json:"authorization,omitempty"``

check SafeAuthorization struct, Type default value is Bearer, so I think `Bear` is fine, but `bearerTokenFile` is also fine
https://github.com/openshift/cluster-monitoring-operator/blob/release-4.11/vendor/github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring/v1/types.go#L1752-L1758
```
type SafeAuthorization struct {
	// Set the authentication type. Defaults to Bearer, Basic will cause an
	// error
	Type string `json:"type,omitempty"`
	// The secret's key that contains the credentials of the request
	Credentials *v1.SecretKeySelector `json:"credentials,omitempty"`
}
```

> Currently supported authentication methods are AWS Signature version 4 (`sigv4`), basic authentication (`basicAuth`),  bearer token (`Bearer`), TLS client (`tlsConfig`), and OAuth 2.0 (`oauth2`). 

Reviewer 4: > I changed all occurrences of Bearer to bearerTokenFile. 

@bburt-rh this is not correct. In the RemoteWrite we have essentially 3 ways of doing authentication:
1. BasicAuth to use this method users have the field `remoteWrite.basicAuth`;
2. Bearer token to use this method users can provide either a `bearerTokenFile` with the field `remoteWrite.bearerTokenFile`  alternatively or specify a secret where the bearer token is located with `remoteWrite.authorization` where `Type: Bearer`.
3. Sigv4 to use this method users have the field `remoteWrite.sigv4`

Hopefully this made it more clear. If you want examples in this link you have test cases for all the methods [1]
[1] https://github.com/openshift/cluster-monitoring-operator/blob/dc7b00300f9a8c0d9cfde08724b9ef58292bab48/pkg/manifests/manifests_test.go#L1271

> Currently supported authentication methods are AWS Signature version 4 (`sigv4`), basic authentication (`basicAuth`),  bearer token (`Bearer`), TLS client (`tlsConfig`), and OAuth 2.0 (`oauth2`). 

Author: I've documented the `Bearer` and `bearerFile` methods of configuring bearer token authorization.

+
* AWS Signature Version 4
+
This method uses AWS Signature Version 4 (`sigv4`) authentication to sign requests.

> This method cannot be used simultaneously with `basic_auth`, `Bearer`, or `oauth2`.

Author: fixed

+
The following shows a `sigv4` sample configuration.

> It assumes you have already created a `sigv4` secret named `sigv4-credentials`in the `openshift-monitoring` namespace.

Reviewer 3: `sigv4-credentials`in 
should be
`sigv4-credentials` in

Author: fixed

Replace the sample `region`, `key`, `profile`, and `roleArn` values with your authentication credentials:
+
[source,yaml]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    prometheusK8s:
      remoteWrite:

>        - url: "https://authorization.remotewrite.com/api/write"

Reviewer 3: ```
      remoteWrite:
       - url: "https://authorization.remotewrite.com/api/write"
```
it is better change to
```
      remoteWrite:
      - url: "https://authorization.remotewrite.com/api/write"
```

        sigv4:
          region: <AWS_region>
          accessKey:
            name: sigv4-credentials
            key: <AWS_access_key>
          secretKey:
            name: sigv4-credentials
            key: <secret>
          profile: <AWS_profile_name>
          roleArn: <AWS_role_arn>
----
+
* basic authentication
+
Basic authorization (`basicAuth`) sets the authorization header on every remote write request with the configured username and password.
+
The following sample shows a `basicAuth` configuration.

> It assumes you have already set up authentication credentials for the endpoint.

Reviewer 3: change to
It assumes you have already set up authentication credentials named `remoteWriteAuth` for the endpoint.

  config.yaml: |
    prometheusK8s:
      remoteWrite:
      - url: "https://remote-write-endpoint.example.com" <1>
        <endpoint_authentication_credentials> <2>
----
+
<1> The URL of the remote write endpoint.
<2> The authentication method and credentials for the endpoint.
Currently supported authentication methods are AWS Signature Version 4, bearer token, basic authentication, OAuth 2.0, and TLS client. 

> See _Supported remote write authentication settings_ below for sample configurations of supported authentication methods.

Reviewer 3: "Supported remote write authentication settings" should add a link to the help part

Author: @juzhao OpenShift documentation standards do not allow me to add a cross-reference link from this module file to another module file. See the "Modules" section of this doc fyi: https://docs.google.com/document/d/1WN7k72PKxcPA__erp5TK-CefmS_sVwE3r40PvpmNk3E/edit#heading=h.k4v49gmkfxi

Reviewer 3: ok, no question

> . Add write relabel configuration values after the authentication credentials:

Reviewer 3: I think this setting is optional for configure remote write storage, but keep it is fine

  config.yaml: |
    prometheusK8s:
      remoteWrite:
      - url: "https://remote-write-endpoint.example.com" <1>
        <endpoint_authentication_credentials> <2>
----
+
<1> The URL of the remote write endpoint.
<2> The authentication method and credentials for the endpoint.

> Currently supported authentication methods are AWS Signature Version 4, bearer token, basic authentication, OAuth 2.0, and TLS client. 

Reviewer 5: ```suggestion
Currently supported authentication methods are AWS Signature Version 4, authentication using HTTP `Authorization` request header, basic authentication, OAuth 2.0, and TLS client. 
```

      - url: "https://remote-write.endpoint"
        basicAuth:
          username:
            name: rw-basic-auth
            key: <username>
          password:
            name: rw-basic-auth
            key: <password>
----
+
* *bearer token* 
+

> A bearer token (`bearerTokenFile`) configuration combines a basic authentication username and password with authorization credentials configured to use a specified token (such as an API key).

Reviewer 3: I think the description is improper, since bearer token sets the `Authorization` header on every scrape request with the configured credentials.

+
The following shows a `bearerTokenFile` authorization sample that uses an authorization token.
It assumes that you have already created a `bearerTokenFile` secret with the name `rw-bearer-auth` in the `openshift-monitoring` namespace.
Substitute your token for the `<token_name>` `key` value. 
+
[source,yaml]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    prometheusK8s:
      remoteWrite:
      - url: "https://authorization.remotewrite.com/api/write"
        authorization:
          type: bearerTokenFile
          credentials:

>             name: rw-berer-auth

Reviewer 3: rw-bearer-auth

Author: fixed

        basicAuth:
          username:

>             name: remoteWriteAuth

Reviewer 3: we should mention what's remoteWriteAuth here.
FYI, for secret, the secret name is improper, since
```
The Secret "remoteWriteAuth" is invalid: metadata.name: Invalid value: "remoteWriteAuth": a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')
```

Author: - changed name from `remoteWriteAuth` to `rw-basic-auth`
- Added a sentence saying `It also assumes that you have already created a `basicAuth` secret with the name `rw-basic-auth` in the `openshift-monitoring` namespace.`

            key: <username>
          password:
            name: remoteWriteAuth
            key: <password>
----
+
* bearer token 
+
A bearer token (`Bearer`) configuration combines a basic authentication username and password with authorization credentials configured to use a specified token (such as an API key).
+
The following shows a `Bearer` authorization sample that uses a `basicAuth` username and password combined with an authorization token.
It assumes you have already created a `Bearer` secret with the name `rw-auth` in the `openshift-monitoring` namespace.
Substitute your username, password, and token for the appropriate `key` values. 
+
[source,yaml]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    prometheusK8s:
      remoteWrite:

>       - url: "https://remote-write.endpoint"

Reviewer 4: Not sure, if in this example, we care about the first two entries of this array since only the last entry contains information actually relevant for user wanting to use this method of configuration

Author: deleted lines 144 to 152 and changed the descriptive text

      - url: "https://basicAuth.remotewrite.com/api/write"
        basicAuth:
          username:
            name: rw-auth
            key: <username>
          password:
            name: rw-auth
            key: <password>
      - url: "https://authorization.remotewrite.com/api/write"

>         authorization:

Reviewer 3: from https://prometheus.io/docs/prometheus/latest/configuration/configuration
I see
```
# Authentication information used to authenticate to the API server.
# Note that `basic_auth`, `authorization` and `oauth2` options are
# mutually exclusive.
```
then we should not configure `basicAuth` and `authorization` together

Reviewer 3: not sure about it, needs confirmation from developer

Author: deleted lines 144 to 152 and changed the descriptive text

          type: Bearer
          credentials:
            name: rw-auth
            key: <token_name>
----
* OAuth 2.0
+

> An OAuth 2.0 (`oauth2`) configuration uses the client credentials grant type. Prometheus fetches an access token from the specified remote write endpoint with the specified client access and secret keys.

Reviewer 2: Prometheus fetches an access token from `tokenUrl`  with the specified client name and client secret, in order to access the remote write endpoint using this token.

Author: Updated the content per your comment

> You cannot use this method simultaneously with `basic_auth`, `Bearer`, or `sigv2` authentication.

Reviewer 3: basic_auth -> basicAuth
sigv2 -> sigv4

Author: fixed

+
The following shows an `oauth2` authentication sample configuration.
It assumes you have already created an `oauth2` secret with the name `oauth2-credentials` in the `openshift-monitoring` namespace.
Replace the sample `key`, `scopes`, and `endpointParams` values with your authentication credentials and other values.
+
[source,yaml]
----
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    prometheusK8s:
      remoteWrite:
        - url: https://test.remotewrite.com/api/write
          remoteTimeout: 30s
          oauth2:
            clientId:
              secret:
                name: oauth2-credentials
                key: <id>
            clientSecret:
              name: oauth2-credentials
              key: <secret>
            tokenUrl: https://example.com/oauth2/token
            scopes:
              - <scope_1>
              - <scope_2>
            endpointParams:
              param1: <value_1>

>               param2: <value_2>

Reviewer 3: it's better to change the format to below, there are additional spaces
```
apiVersion: v1
kind: ConfigMap
metadata:
  name: cluster-monitoring-config
  namespace: openshift-monitoring
data:
  config.yaml: |
    prometheusK8s:
      remoteWrite:
      - url: https://test.remotewrite.com/api/write
        remoteTimeout: 30s
        oauth2:
          clientId:
            secret:
              name: oauth2-credentials
              key: <id>
          clientSecret:
            name: oauth2-credentials
            key: <secret>
          tokenUrl: https://example.com/oauth2/token
          scopes:
          - <scope_1>
          - <scope_2>
          endpointParams:
            param1: <value_1>
            param2: <value_2>
```

Author: fixed

Reviewer 3: see format is still wrong, please see my example

----
+
* TLS client
+
A TLS client (`tlsConfig`) configuration specifies the CA certificate, the client certificate, and the client key file information used to authenticate with the remote write endpoint server using TLS.
+
The following sample shows a `tlsConfig` configuration.

> It assumes that you have already created a CA certificate file, a client certificate file, and a client key file.

Reviewer 6: It assumes you have already created a CA certificate file, a client certificate file, and a client key file. You also have created a `tls` secret with the name `selfsigned-mtls-bundle` in the `openshift-monitoring` namespace.

Author: Added the new sentence

          ca:
            secret:
              name: selfsigned-mtls-bundle
              key: <ca.crt>
          cert:
            secret:
              name: selfsigned-mtls-bundle
              key: <client.crt>
          keySecret:
            name: selfsigned-mtls-bundle

>             key: <client.key>

Reviewer 3: see from https://deploy-preview-44713--osdocs.netlify.app/openshift-enterprise/latest/monitoring/configuring-the-monitoring-stack.html#configuring_remote_write_storage_configuring-the-monitoring-stack
the `tlsConfig` configmap format is different with others

Author: fixed
